analyzer/emerald/incoming.go

```diff
@@ -0,0 +1,365 @@
+	return addr, nil
+}
+
+//nolint:unparam
+func registerRelatedSdkAddress(relatedAddresses map[string]bool, sdkAddr *sdkTypes.Address) (string, error) {
+	addr, err := common.StringifySdkAddress(sdkAddr)
+	if err != nil {
+		return "", err
+	}
+
+	relatedAddresses[addr] = true
+
+	return addr, nil
+}
+
+func registerRelatedAddressSpec(addressPreimages map[string]*AddressPreimageData, relatedAddresses map[string]bool, as *sdkTypes.AddressSpec) (string, error) {
+	addr, err := registerAddressSpec(addressPreimages, as)
+	if err != nil {
+		return "", err
+	}
+
+	relatedAddresses[addr] = true
+
+	return addr, nil
+}
+
+//nolint:unparam
+func registerRelatedEthAddress(addressPreimages map[string]*AddressPreimageData, relatedAddresses map[string]bool, ethAddr []byte) (string, error) {
+	addr, err := registerEthAddress(addressPreimages, ethAddr)
+	if err != nil {
+		return "", err
+	}
+
+	relatedAddresses[addr] = true
+
+	return addr, nil
+}
+
+//nolint:gocyclo
+func extractRound(sigContext signature.Context, b *block.Block, txrs []*sdkClient.TransactionWithResults) (*BlockData, error) {
+	var blockData BlockData
+	blockData.Hash = b.Header.EncodedHash().String()
+	blockData.NumTransactions = len(txrs)
+	blockData.TransactionData = make([]*BlockTransactionData, 0, len(txrs))
+	blockData.AddressPreimages = map[string]*AddressPreimageData{}
+	for i, txr := range txrs {
+		// fmt.Printf("%#v\n", txr)
+		var blockTransactionData BlockTransactionData
+		blockTransactionData.Index = i
+		blockTransactionData.Hash = txr.Tx.Hash().Hex()
+		if len(txr.Tx.AuthProofs) == 1 && txr.Tx.AuthProofs[0].Module == "evm.ethereum.v0" {
+			ethHash := hex.EncodeToString(common.Keccak256(txr.Tx.Body))
+			blockTransactionData.EthHash = &ethHash
+		}
+		blockTransactionData.RelatedAccountAddresses = map[string]bool{}
+		tx, err := common.VerifyUtx(sigContext, &txr.Tx)
+		if err != nil {
+			err = fmt.Errorf("tx %d: %w", i, err)
+			fmt.Println(err)
+			tx = nil
+		}
+		if tx != nil {
+			blockTransactionData.SignerData = make([]*BlockTransactionSignerData, 0, len(tx.AuthInfo.SignerInfo))
+			for j, si := range tx.AuthInfo.SignerInfo {
+				var blockTransactionSignerData BlockTransactionSignerData
+				blockTransactionSignerData.Index = j
+				addr, err1 := registerRelatedAddressSpec(blockData.AddressPreimages, blockTransactionData.RelatedAccountAddresses, &si.AddressSpec)
+				if err1 != nil {
+					return nil, fmt.Errorf("tx %d signer %d visit address spec: %w", i, j, err1)
+				}
+				blockTransactionSignerData.Address = addr
+				blockTransactionSignerData.Nonce = int(si.Nonce)
+				blockTransactionData.SignerData = append(blockTransactionData.SignerData, &blockTransactionSignerData)
+			}
+			if err = common.VisitCall(&tx.Call, &txr.Result, &common.CallHandler{
+				AccountsTransfer: func(body *accounts.Transfer) error {
+					if _, err = registerRelatedSdkAddress(blockTransactionData.RelatedAccountAddresses, &body.To); err != nil {
+						return fmt.Errorf("to: %w", err)
```

Comment:
Can you make the error message more descriptive please? Here and in a few other places.

Reply:
other layers of wrapping add more description, e.g. that it's trying to marshal an address, that it's doing the nth transaction of the mth block